### DISCUSSION TITLE
async selectAtom may stop propagating values

### DISCUSSION THREAD

**Discussion by satoren** [pmndrs/jotai#2195] [Bug report]
If the baseAtom is updated before the async selectAtom resolve, the selectAtom's will forever stop propagating values.

https://codesandbox.io/s/dank-bird-p96jqw?file=/src/App.tsx

The selectAtom resolves after 500ms, but if you press the button to update the baseAtom at shorter intervals than that, the selectAtom will no longer return a value.

This does not occur with simple async derived atom
https://codesandbox.io/s/shy-waterfall-ysz6pr?file=/src/App.tsx

**Comment by dai-shi** [comment]
Do you think it's the same or related with https://github.com/pmndrs/jotai/discussions/2192#discussioncomment-7314439 ?

**Reply by satoren** [reply]
It looks like it. I will close this and try again after it is resolved, and reopen it if there is a problem.

**Comment by dmaskasky** [comment]
That is my suspicion

**Comment by satoren** [comment]
This was not fixed in v2.5.0.

https://codesandbox.io/s/shy-waterfall-ysz6pr?file=/src/App.tsx

**Reply by dai-shi** [reply]
:cry:

**Reply by dai-shi** [reply]
I'm not sure if it's fixable. Looking at the code again, I think a selector returning a promise is never supported. We should probably note that in docs.

I think the workaround is unwrap first and the selectAtom.

Please note that selectAtom is a hack by design and it should only be used for escape hatches. (It means the data flow isn't pure/declarative. So whenever you can, prefer derived atoms.)

That said, anyone is welcome to investigate it further.

**Reply by satoren** [reply]
https://codesandbox.io/s/headless-worker-dnvf3f?file=/src/selectAtom.ts

Removing the Promise code appears to be working correctly. (However, in the case of Promise, the `equalityFn` would be practically useless)

**Reply by dai-shi** [reply]
Yeah, I think not supporting async selector sounds reasonable.
